src/MISC/fix_ipi.cpp

```diff
@@ -362,11 +368,39 @@ void FixIPI::initial_integrate(int /*vflag*/)
   // ensure atoms are in current box & update box via shrink-wrap
   // has to be be done before invoking Irregular::migrate_atoms()
   //   since it requires atoms be inside simulation box
-
-  if (domain->triclinic) domain->x2lamda(atom->nlocal);
-  domain->pbc();
-  domain->reset_box();
-  if (domain->triclinic) domain->lamda2x(atom->nlocal);
+  if (neighbor->ncalls == 0) {
+    // just fold coordinates at the first step
+    if (domain->triclinic) domain->x2lamda(atom->nlocal);
+    domain->pbc();
+    domain->reset_box();
+    if (domain->triclinic) domain->lamda2x(atom->nlocal);
+  } else {
+    // "unwraps" the trajectory because we have no guarantee of what has
+    // happened server-side to the atoms folding, and we want to have continuous
+    // trajectories to build NL in a meaningful way and as rarely as possible
+
+    auto xhold = neighbor->get_xhold();
+    for (int i = 0; i < nlocal; i++) {
+      if (mask[i] & groupbit) {
+        x[i][0] -= xhold[i][0];
+        x[i][1] -= xhold[i][1];
+        x[i][2] -= xhold[i][2];
+      }
+    }
+    // applies PBC to the displacements
+    if (domain->triclinic) domain->x2lamda(atom->nlocal);
+    domain->pbc();
+    domain->reset_box();
+    if (domain->triclinic) domain->lamda2x(atom->nlocal);
+    // recovers "unwrapped" positions
+    for (int i = 0; i < nlocal; i++) {
+      if (mask[i] & groupbit) {
+        x[i][0] += xhold[i][0];
+        x[i][1] += xhold[i][1];
+        x[i][2] += xhold[i][2];
+      }
+    }
+  }
```

Comment:
I could just be missing something, but I don't understand how this scheme works for a general box. Say the box doesn't contain the origin $\vec{0}$. If a particle doesn't move, such that $\vec{x} = \vec{x}\_{hold}$, then `domain->pbc()` would shift $(\vec{x} - \vec{x}\_{hold}) = \vec{0}$ across periodic images back into the box. Adding $\vec{x}\_{hold}$ again would then shift the particle back out of the box and trigger an unnecessary neighbor rebuild since now $\vec{x} \ne \vec{x}\_{hold}$.

If I understand the intent correctly, what about alternatively shifting `xhold` to the nearest periodic image, e.g. something like
```
if (domain->triclinic) domain->x2lamda(atom->nlocal);
domain->pbc();
domain->reset_box();
if (domain->triclinic) domain->lamda2x(atom->nlocal);
if (neigh->ncalls != 0) {
  for (int i = 0; i < nlocal; i++) {
    delx = xhold[i][0] - x[i][0];
    ...
    domain->minimum_image(delx, dely, delz);
    xhold[i][0] = x[i][0] + delx;
    ...
  }
}
```

Reply:
This works because `xhold` is shifted back to the 000 replica when the neighbor list is updated. So this code makes sure the displacement is has a minimum image convention applied when it's computed in the neighbor list code, without touching that part of the code that is too core to be messed with, IMO.

Reply:
I see. So what happens if $\vec{x} = \vec{x}\_{hold}$ (both in the same image) but the box doesn't contain the origin? Does the neighbor list get rebuilt?

Since `domain->pbc()` acts on position vectors and its behavior depends on the location of the `boxlo`, I would not assume feeding it a displacement vector (invariant to origin translations) would produce the desired behavior for all box geometries and displacements.

Reply:
Thanks, I agree this is cleaner. I have reformulated the code in `fix_ipi.cpp` - only difference is I am modifying the coordinates rather than xhold, again following the principle of minimal messing with the core classes.

Reply:
Cool, happy to help. Hopefully that works to improve performance!